tidy3d/plugins/smatrix/component_modelers/terminal.py

```diff
@@ -507,5 +517,126 @@ def get_radiation_monitor_by_name(self, monitor_name: str) -> DirectivityMonitor
+                extrude_to = back_pec_plane.center[inj_axis]
+
+                # move cutting plane beyond the waveport plane along the `ModeSource` injection direction.
+                center = list(back_pec_plane.center)
+                center[inj_axis] = port.center[inj_axis] - sign * fp_eps * box.size[inj_axis]
+                cutting_plane = back_pec_plane.updated_copy(center=center)
+
+                # define extrusion bounds
+                extrusion_bounds = [cutting_plane.center[inj_axis], extrude_to][::sign]
+
+                # loop over structures and extrude those that intersect a waveport plane
+                for structure in sim.structures:
+                    # get geometries that intersect the plane on which the waveport is defined
+
+                    shapely_geom = cutting_plane.intersections_with(structure.geometry)
+
+                    polygon_list = []
+                    for geom in shapely_geom:
+                        polygon_list = polygon_list + ClipOperation.to_polygon_list(geom)
+
+                    new_geoms = []
+                    # loop over identified geometries and extrude them
+                    for polygon in polygon_list:
+                        # construct outer shell of an extruded geometry first
+                        exterior_vertices = np.array(polygon.exterior.coords)
+                        outer_shell = PolySlab(
+                            axis=inj_axis, slab_bounds=extrusion_bounds, vertices=exterior_vertices
+                        )
+
+                        # construct innner shells that represent holes
+                        hole_polyslabs = [
+                            PolySlab(
+                                axis=inj_axis,
+                                slab_bounds=extrusion_bounds,
+                                vertices=np.array(hole.coords),
+                            )
+                            for hole in polygon.interiors
+                        ]
+
+                        # construct final geometry by removing inner holes from outer shell
+                        if hole_polyslabs:
+                            holes = GeometryGroup(geometries=hole_polyslabs)
+                            extruded_slab_new = ClipOperation(
+                                operation="difference", geometry_a=outer_shell, geometry_b=holes
+                            )
+                        else:
+                            extruded_slab_new = outer_shell
+
+                        # append extruded geometry
+                        new_geoms.append(extruded_slab_new)
+
+                    # if current port does not intersect any structures raise error
+                    if not new_geoms:
+                        raise SetupError(
+                            f"The 'WavePort' '{port.name}' does not intersect any structures."
+                            f"Please ensure that it is located within or at the boundary of a structure."
+                        )
```

Comment:
this would error if the port intersects some structures, but not all structures, right? You should probably just skip non-intersecting structures (`len(shapely_geom) == 0`) and check `new_structures` in the end instead